From 510c380409fdf1d5748f18f5394c5d7db18ccf07 Mon Sep 17 00:00:00 2001
From: Rob Clark <rob@ti.com>
Date: Mon, 24 May 2010 16:49:20 -0500
Subject: [PATCH 4/6] Add GstEventVstab event

This event can be used to set tilt/pan to take effect on the following buffer.
---
 gst/gstevent.c |   50 ++++++++++++++++++++++++++++++++++++++++++++++++++
 gst/gstevent.h |    8 ++++++++
 gst/gstquark.c |    2 +-
 gst/gstquark.h |    5 ++++-
 4 files changed, 63 insertions(+), 2 deletions(-)

diff --git a/gst/gstevent.c b/gst/gstevent.c
index 3c1b0b9..69b57ec 100644
--- a/gst/gstevent.c
+++ b/gst/gstevent.c
@@ -116,6 +116,7 @@ static GstEventQuarks event_quarks[] = {
   {GST_EVENT_TAG, "tag", 0},
   {GST_EVENT_BUFFERSIZE, "buffersize", 0},
   {GST_EVENT_SINK_MESSAGE, "sink-message", 0},
+  {GST_EVENT_VSTAB, "vstab", 0},
   {GST_EVENT_QOS, "qos", 0},
   {GST_EVENT_SEEK, "seek", 0},
   {GST_EVENT_NAVIGATION, "navigation", 0},
@@ -1236,3 +1237,52 @@ gst_event_parse_sink_message (GstEvent * event, GstMessage ** msg)
         GST_MESSAGE (gst_value_dup_mini_object (gst_structure_id_get_value
             (structure, GST_QUARK (MESSAGE))));
 }
+
+/**
+ * gst_event_new_vstab:
+ * @top:  the new offset to top of sub-image
+ * @left:  the new offset to left of sub-image
+ *
+ * Create a new vstab event.
+ */
+GstEvent *
+gst_event_new_vstab (gint top, gint left)
+{
+  GstEvent *event;
+  GstStructure *structure;
+
+  GST_CAT_INFO (GST_CAT_EVENT, "creating vstab event: top=%d, left=%d",
+      top, left);
+
+  structure = gst_structure_id_new (GST_QUARK (EVENT_VSTAB),
+      GST_QUARK (TOP), G_TYPE_INT, top,
+      GST_QUARK (LEFT), G_TYPE_INT, left, NULL);
+  event = gst_event_new_custom (GST_EVENT_VSTAB, structure);
+
+  return event;
+}
+
+/**
+ * gst_event_parse_vstab:
+ * @event: The event to query
+ * @top: A pointer to store top offset in
+ * @left: A pointer to store left offset in
+ *
+ * Parse the vstab event.
+ */
+void
+gst_event_parse_vstab (GstEvent * event, gint * top, gint * left)
+{
+  const GstStructure *structure;
+
+  g_return_if_fail (GST_IS_EVENT (event));
+  g_return_if_fail (GST_EVENT_TYPE (event) == GST_EVENT_VSTAB);
+
+  structure = gst_event_get_structure (event);
+  if (top)
+    *top = g_value_get_int (gst_structure_id_get_value (structure,
+            GST_QUARK (TOP)));
+  if (left)
+    *left = g_value_get_int (gst_structure_id_get_value (structure,
+            GST_QUARK (LEFT)));
+}
diff --git a/gst/gstevent.h b/gst/gstevent.h
index 6fc809f..3402ccb 100644
--- a/gst/gstevent.h
+++ b/gst/gstevent.h
@@ -93,6 +93,9 @@ typedef enum {
  * @GST_EVENT_SINK_MESSAGE: An event that sinks turn into a message. Used to
  *                          send messages that should be emitted in sync with
  *                          rendering.
+ * @GST_EVENT_VSTAB: An event that can set horizontal (pan/scan) and vertical
+ *                   (tilt/scan) offset within a larger image.  This event
+ *                   precedes the buffer to which it applies.
  * @GST_EVENT_QOS: A quality message. Used to indicate to upstream elements
  *                 that the downstream elements are being starved of or
  *                 flooded with data.
@@ -133,6 +136,7 @@ typedef enum {
   GST_EVENT_TAG                   = GST_EVENT_MAKE_TYPE (7, FLAG(DOWNSTREAM) | FLAG(SERIALIZED)),
   GST_EVENT_BUFFERSIZE            = GST_EVENT_MAKE_TYPE (8, FLAG(DOWNSTREAM) | FLAG(SERIALIZED)),
   GST_EVENT_SINK_MESSAGE          = GST_EVENT_MAKE_TYPE (9, FLAG(DOWNSTREAM) | FLAG(SERIALIZED)),
+  GST_EVENT_VSTAB                 = GST_EVENT_MAKE_TYPE (10, FLAG(DOWNSTREAM) | FLAG(SERIALIZED)),
   /* upstream events */
   GST_EVENT_QOS                   = GST_EVENT_MAKE_TYPE (15, FLAG(UPSTREAM)),
   GST_EVENT_SEEK                  = GST_EVENT_MAKE_TYPE (16, FLAG(UPSTREAM)),
@@ -483,6 +487,10 @@ GstEvent*       gst_event_new_step              (GstFormat format, guint64 amoun
 void            gst_event_parse_step            (GstEvent *event, GstFormat *format, guint64 *amount,
                                                  gdouble *rate, gboolean *flush, gboolean *intermediate);
 
+/* vstab event */
+GstEvent *      gst_event_new_vstab             (gint top, gint left);
+void            gst_event_parse_vstab           (GstEvent * event, gint * top, gint * left);
+
 G_END_DECLS
 
 #endif /* __GST_EVENT_H__ */
diff --git a/gst/gstquark.c b/gst/gstquark.c
index 58badca..491b1d9 100644
--- a/gst/gstquark.c
+++ b/gst/gstquark.c
@@ -50,7 +50,7 @@ static const gchar *_quark_strings[] = {
   "intermediate", "GstMessageStepStart", "active", "eos", "sink-message",
   "message", "GstMessageQOS", "running-time", "stream-time", "jitter",
   "quality", "processed", "dropped", "buffering-ranges", "GstQueryBuffers",
-  "caps", "count", "width", "height"
+  "caps", "count", "width", "height", "GstEventVstab", "top", "left"
 };
 
 GQuark _priv_gst_quark_table[GST_QUARK_MAX];
diff --git a/gst/gstquark.h b/gst/gstquark.h
index f4c8e0f..506cf95 100644
--- a/gst/gstquark.h
+++ b/gst/gstquark.h
@@ -132,8 +132,11 @@ typedef enum _GstQuarkId
   GST_QUARK_COUNT = 103,
   GST_QUARK_WIDTH = 104,
   GST_QUARK_HEIGHT = 105,
+  GST_QUARK_EVENT_VSTAB = 106,
+  GST_QUARK_TOP = 107,
+  GST_QUARK_LEFT = 108,
 
-  GST_QUARK_MAX = 106
+  GST_QUARK_MAX = 109
 } GstQuarkId;
 
 extern GQuark _priv_gst_quark_table[GST_QUARK_MAX];
-- 
1.7.1

